7_drivers_munge/src/GCM_driver_utils.R

```diff
@@ -242,26 +245,125 @@ download_gcm_data <- function(out_file_template, query_geom,
-  )
-  wait(gcm_job)
+    wait(gcm_job)
+    gcm_job_status <- check(gcm_job)$statusType
+  },
+  # Check the value of the last thing in the expr above (denoted
+  # by the ".") to decide if you should retry or not
+  until = ~ . == "ProcessSucceeded",
+  max_tries = 5)
+
   my_data <- result(gcm_job, with.units = TRUE)
 
   # Build out_file name
-  out_file <- sprintf(out_file_template, gcm_name, query_geom$tile_no[1])
+  out_file <- sprintf(out_file_template,
+                      gcm_name,
+                      gcm_projection_period,
+                      query_geom$tile_no[1])
 
   # write file
   arrow::write_feather(my_data, out_file)
 
   return(out_file)
 }
+
+#' @title Pull out individual file information from a dynamically mapped target
+#' @description Use a dynamically mapped target to extract information about
+#' the individual branch files to build a table that can group files and then
+#' trigger rebuilds when individual files within a group change.
+#' @param dynamic_branch_names character vector of the names of each branch created
+#' from running `names()` using the target dynamically branched target as input.
+#' @return a tibble with three columns (`name` = target name of the branch, `path`
+#' = character string of the file created by that branch, and `data` = hash code
+#' indicating the current state of the file) and a row for each branch.
+build_branch_file_hash_table <- function(dynamic_branch_names) {
+  # Get metadata for all the branches of the dynamic target
+  tar_meta(all_of(dynamic_branch_names), c("path", "data")) %>%
+    # There is just one path per branch, so remove the `list` format from this column
+    unnest(path)
+}
+
+#' @title Summarize hourly output from geoknife to daily
+#' @description the final GCM driver data will need to be daily, but geoknife
+#' returns hourly values. This step summarizes the data into daily values. It
+#' creates a file with the exact same name, except that the "_raw" part of the
+#' `in_file` filepath is replaced with "_daily".
+#' @param in_file filepath to a feather file containing the hourly geoknife data
+munge_to_glm <- function(in_file) {
+
+  daily_data <- arrow::read_feather(in_file) %>%
+
+    # Pivot to long format first to get cells as a column.
+    pivot_longer(cols = -c(DateTime, variable, statistic, units),
+                 names_to = "cell", values_to = "val") %>%
+    mutate(cell = as.numeric(cell)) %>%
+
+    # Now pivot wider to makes each variable a column for easy, readable munging
+    # Also, removes `units` and `statistic` columns which are specific to each variable
+    pivot_wider(id_cols = c("DateTime", "cell"), names_from = variable, values_from = val) %>%
+
+    # Unit conversions to get GLM-ready variables from GDP ones
+    mutate(AirTemp = tas - 273.15, # Convert GDP air temperature (tas) from Kelvin to Celcius
+           Rain = pr # TODO: Convert GDP precipitation (pr) from kg/m2/s to meters/hour
```

Comment:
@jread-usgs interested in what your thoughts are for this variable.

Reply:
Ahh, I looked at Hayley's issue #15 before this so expect some duplicate info here :)

For this conversion, you can assume the density of water is 1000kg/m3 and use that to calculate the flux in the GLM units.

Reply:
And note we'll actually need units of meters/day, not meters/hour. I just did this to check the GLM results, and that conversion comes out to `m/day = kg m-2 s-1 * 86.4`

Here's how I got to that so that you can double check it and make sure it seems right:

kg m-2 s-1 = density of water (kg/m3) * rate of rainfall (m/s)

Since we want the rate of rainfall, we divide the GDP value in kg m-2 s-1 by the density of water (1000 kg/m3), and then multiply by 86400 (60 secs * 60 min * 24 hrs) to convert the resultant rate (m/s) to m/day. So that results in a conversion factor of 86.4.

Reply:
Nice work, Hayley. One thing to add here is that the _GDP_ values aren't always going to be in the same units (different data sources on GDP have different units) but we can assume the GCMs have the same units. An important check we can use on this conversions is to verify the units column in the GDP result is what we're expecting. That is one good reason to bring along that optional column output. That way if we point this to a slightly different dataset/variable in the future, we'd fail early and we'd know why if the units differed.

Reply:
Ahh OK, so we are getting the _rate_ into meters/day (even though we will have a different daily rate per hourly timestep after this `mutate()`, then we will average those daily rainfall rates to get a single daily rainfall rate per day in `summarize()` below. So, my comment about summing per day is not what we want.